framework/ChiMesh/VolumeMesher/chi_volumemesher_01c_set_IDs.cc

```diff
@@ -109,4 +110,223 @@ SetMatIDToAll(int mat_id)
   chi::log.Log()
     << chi::program_timer.GetTimeString()
     << " Done setting material id " << mat_id << " to all cells";
+}
+
+
+//###################################################################
+/**Sets material id's using a lua function. The lua function is called
+with for each cell with 4 arguments, the cell's centroid x,y,z values
+and the cell's current material id.
+
+The lua function's prototype should be:
+\code
+function LuaFuncName(x,y,z,id)
+ --stuff
+end
+\endcode
+*/
+void chi_mesh::VolumeMesher::
+  SetMatIDFromLuaFunction(const std::string& lua_fname)
+{
+  const std::string fname = "chi_mesh::VolumeMesher::SetMatIDFromLuaFunction";
+
+  chi::log.Log0Verbose1()
+    << chi::program_timer.GetTimeString()
+    << " Setting material id from lua function.";
+
+  //============================================= Define console call
+  auto L = chi::console.consoleState;
+  auto CallLuaXYZFunction = [&L,&lua_fname,&fname](const chi_mesh::Cell& cell)
+  {
+    //============= Load lua function
+    lua_getglobal(L, lua_fname.c_str());
+
+    //============= Error check lua function
+    if (not lua_isfunction(L, -1))
+      throw std::logic_error(fname + " attempted to access lua-function, " +
+                               lua_fname + ", but it seems the function"
+                                             " could not be retrieved.");
+
+    const auto& xyz = cell.centroid;
+
+    //============= Push arguments
+    lua_pushnumber(L, xyz.x);
+    lua_pushnumber(L, xyz.y);
+    lua_pushnumber(L, xyz.z);
+    lua_pushinteger(L, cell.material_id);
+
+    //============= Call lua function
+    //4 arguments, 1 result (double), 0=original error object
+    int lua_return;
+    if (lua_pcall(L,4,1,0) == 0)
+    {
+      LuaCheckNumberValue(fname, L, -1);
```

Comment:
this fname is what I was missing in my lua calls in CFEM/DFEM in order not to hard-code the funct name. Now I know how to improve those modules. thanks